<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title></title>
  <meta name="description" content="">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  


<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="MacDonald_Annotation.html"><a href="#introduction-to-bioconductor-annotation-resources"><i class="fa fa-check"></i><b>1</b> Introduction to Bioconductor annotation resources</a><ul>
<li class="chapter" data-level="1.1" data-path="MacDonald_Annotation.html"><a href="#instructors"><i class="fa fa-check"></i><b>1.1</b> Instructors</a></li>
<li class="chapter" data-level="1.2" data-path="MacDonald_Annotation.html"><a href="#workshop-description"><i class="fa fa-check"></i><b>1.2</b> Workshop Description</a><ul>
<li class="chapter" data-level="1.2.1" data-path="MacDonald_Annotation.html"><a href="#prerequisites"><i class="fa fa-check"></i><b>1.2.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2.2" data-path="MacDonald_Annotation.html"><a href="#workshop-participation"><i class="fa fa-check"></i><b>1.2.2</b> Workshop Participation</a></li>
<li class="chapter" data-level="1.2.3" data-path="MacDonald_Annotation.html"><a href="#r-bioconductor-packages-used"><i class="fa fa-check"></i><b>1.2.3</b> <em>R</em> / <em>Bioconductor</em> packages used</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="MacDonald_Annotation.html"><a href="#workshop-goals-and-objectives"><i class="fa fa-check"></i><b>1.3</b> Workshop goals and objectives</a><ul>
<li class="chapter" data-level="1.3.1" data-path="MacDonald_Annotation.html"><a href="#learning-goals"><i class="fa fa-check"></i><b>1.3.1</b> Learning goals</a></li>
<li class="chapter" data-level="1.3.2" data-path="MacDonald_Annotation.html"><a href="#learning-objectives"><i class="fa fa-check"></i><b>1.3.2</b> Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="MacDonald_Annotation.html"><a href="#annotation-workshop"><i class="fa fa-check"></i><b>1.4</b> Annotation Workshop</a><ul>
<li class="chapter" data-level="1.4.1" data-path="MacDonald_Annotation.html"><a href="#goals-for-this-workshop"><i class="fa fa-check"></i><b>1.4.1</b> Goals for this workshop</a></li>
<li class="chapter" data-level="1.4.2" data-path="MacDonald_Annotation.html"><a href="#what-do-we-mean-by-annotation"><i class="fa fa-check"></i><b>1.4.2</b> What do we mean by annotation?</a></li>
<li class="chapter" data-level="1.4.3" data-path="MacDonald_Annotation.html"><a href="#specific-goal"><i class="fa fa-check"></i><b>1.4.3</b> Specific goal</a></li>
<li class="chapter" data-level="1.4.4" data-path="MacDonald_Annotation.html"><a href="#data-containers"><i class="fa fa-check"></i><b>1.4.4</b> Data containers</a></li>
<li class="chapter" data-level="1.4.5" data-path="MacDonald_Annotation.html"><a href="#expressionset"><i class="fa fa-check"></i><b>1.4.5</b> ExpressionSet</a></li>
<li class="chapter" data-level="1.4.6" data-path="MacDonald_Annotation.html"><a href="#expressionset-continued"><i class="fa fa-check"></i><b>1.4.6</b> ExpressionSet (continued)</a></li>
<li class="chapter" data-level="1.4.7" data-path="MacDonald_Annotation.html"><a href="#expressionset-continued-1"><i class="fa fa-check"></i><b>1.4.7</b> ExpressionSet (continued)</a></li>
<li class="chapter" data-level="1.4.8" data-path="MacDonald_Annotation.html"><a href="#bioc-containers-vs-basic-structures"><i class="fa fa-check"></i><b>1.4.8</b> BioC containers vs basic structures</a></li>
<li class="chapter" data-level="1.4.9" data-path="MacDonald_Annotation.html"><a href="#annotation-sources"><i class="fa fa-check"></i><b>1.4.9</b> Annotation sources</a></li>
<li class="chapter" data-level="1.4.10" data-path="MacDonald_Annotation.html"><a href="#interacting-with-annodb-packages"><i class="fa fa-check"></i><b>1.4.10</b> Interacting with AnnoDb packages</a></li>
<li class="chapter" data-level="1.4.11" data-path="MacDonald_Annotation.html"><a href="#simple-example"><i class="fa fa-check"></i><b>1.4.11</b> Simple example</a></li>
<li class="chapter" data-level="1.4.12" data-path="MacDonald_Annotation.html"><a href="#questions"><i class="fa fa-check"></i><b>1.4.12</b> Questions!</a></li>
<li class="chapter" data-level="1.4.13" data-path="MacDonald_Annotation.html"><a href="#more-questions"><i class="fa fa-check"></i><b>1.4.13</b> More questions!</a></li>
<li class="chapter" data-level="1.4.14" data-path="MacDonald_Annotation.html"><a href="#another-example"><i class="fa fa-check"></i><b>1.4.14</b> Another example</a></li>
<li class="chapter" data-level="1.4.15" data-path="MacDonald_Annotation.html"><a href="#the-mapids-function"><i class="fa fa-check"></i><b>1.4.15</b> The <code>mapIds</code> function</a></li>
<li class="chapter" data-level="1.4.16" data-path="MacDonald_Annotation.html"><a href="#choices-for-multivals"><i class="fa fa-check"></i><b>1.4.16</b> Choices for multiVals</a></li>
<li class="chapter" data-level="1.4.17" data-path="MacDonald_Annotation.html"><a href="#choices-for-multivals-continued"><i class="fa fa-check"></i><b>1.4.17</b> Choices for multiVals (continued)</a></li>
<li class="chapter" data-level="1.4.18" data-path="MacDonald_Annotation.html"><a href="#chipdborgdb-questions"><i class="fa fa-check"></i><b>1.4.18</b> ChipDb/OrgDb questions</a></li>
<li class="chapter" data-level="1.4.19" data-path="MacDonald_Annotation.html"><a href="#txdb-packages"><i class="fa fa-check"></i><b>1.4.19</b> TxDb packages</a></li>
<li class="chapter" data-level="1.4.20" data-path="MacDonald_Annotation.html"><a href="#ensdb-packages"><i class="fa fa-check"></i><b>1.4.20</b> EnsDb packages</a></li>
<li class="chapter" data-level="1.4.21" data-path="MacDonald_Annotation.html"><a href="#transcript-packages"><i class="fa fa-check"></i><b>1.4.21</b> Transcript packages</a></li>
<li class="chapter" data-level="1.4.22" data-path="MacDonald_Annotation.html"><a href="#granges"><i class="fa fa-check"></i><b>1.4.22</b> GRanges</a></li>
<li class="chapter" data-level="1.4.23" data-path="MacDonald_Annotation.html"><a href="#grangeslist"><i class="fa fa-check"></i><b>1.4.23</b> GRangesList</a></li>
<li class="chapter" data-level="1.4.24" data-path="MacDonald_Annotation.html"><a href="#other-accessors"><i class="fa fa-check"></i><b>1.4.24</b> Other accessors</a></li>
<li class="chapter" data-level="1.4.25" data-path="MacDonald_Annotation.html"><a href="#why-ranges-objects"><i class="fa fa-check"></i><b>1.4.25</b> Why *Ranges objects</a></li>
<li class="chapter" data-level="1.4.26" data-path="MacDonald_Annotation.html"><a href="#ranges-use-cases"><i class="fa fa-check"></i><b>1.4.26</b> *Ranges use cases</a></li>
<li class="chapter" data-level="1.4.27" data-path="MacDonald_Annotation.html"><a href="#summarizedexperiment-objects"><i class="fa fa-check"></i><b>1.4.27</b> SummarizedExperiment objects</a></li>
<li class="chapter" data-level="1.4.28" data-path="MacDonald_Annotation.html"><a href="#txdb-exercises"><i class="fa fa-check"></i><b>1.4.28</b> TxDb exercises</a></li>
<li class="chapter" data-level="1.4.29" data-path="MacDonald_Annotation.html"><a href="#organismdb-packages"><i class="fa fa-check"></i><b>1.4.29</b> OrganismDb packages</a></li>
<li class="chapter" data-level="1.4.30" data-path="MacDonald_Annotation.html"><a href="#organismdb-packages-1"><i class="fa fa-check"></i><b>1.4.30</b> OrganismDb packages</a></li>
<li class="chapter" data-level="1.4.31" data-path="MacDonald_Annotation.html"><a href="#organismdb-exercises"><i class="fa fa-check"></i><b>1.4.31</b> OrganismDb exercises</a></li>
<li class="chapter" data-level="1.4.32" data-path="MacDonald_Annotation.html"><a href="#organism.dplyr-package"><i class="fa fa-check"></i><b>1.4.32</b> Organism.dplyr package</a></li>
<li class="chapter" data-level="1.4.33" data-path="MacDonald_Annotation.html"><a href="#organism.dply-example"><i class="fa fa-check"></i><b>1.4.33</b> Organism.dply example</a></li>
<li class="chapter" data-level="1.4.34" data-path="MacDonald_Annotation.html"><a href="#organism.dplyr-exercises"><i class="fa fa-check"></i><b>1.4.34</b> Organism.dplyr exercises</a></li>
<li class="chapter" data-level="1.4.35" data-path="MacDonald_Annotation.html"><a href="#bsgenome-packages"><i class="fa fa-check"></i><b>1.4.35</b> BSgenome packages</a></li>
<li class="chapter" data-level="1.4.36" data-path="MacDonald_Annotation.html"><a href="#bsgenome-packages-1"><i class="fa fa-check"></i><b>1.4.36</b> BSgenome packages</a></li>
<li class="chapter" data-level="1.4.37" data-path="MacDonald_Annotation.html"><a href="#bsgenome-packages-2"><i class="fa fa-check"></i><b>1.4.37</b> BSgenome packages</a></li>
<li class="chapter" data-level="1.4.38" data-path="MacDonald_Annotation.html"><a href="#bsgenome-exercises"><i class="fa fa-check"></i><b>1.4.38</b> BSgenome exercises</a></li>
<li class="chapter" data-level="1.4.39" data-path="MacDonald_Annotation.html"><a href="#annotationhub"><i class="fa fa-check"></i><b>1.4.39</b> AnnotationHub</a></li>
<li class="chapter" data-level="1.4.40" data-path="MacDonald_Annotation.html"><a href="#querying-annotationhub"><i class="fa fa-check"></i><b>1.4.40</b> Querying AnnotationHub</a></li>
<li class="chapter" data-level="1.4.41" data-path="MacDonald_Annotation.html"><a href="#annotationhub-data-providers"><i class="fa fa-check"></i><b>1.4.41</b> AnnotationHub Data providers</a></li>
<li class="chapter" data-level="1.4.42" data-path="MacDonald_Annotation.html"><a href="#annotationhub-data-classes"><i class="fa fa-check"></i><b>1.4.42</b> AnnotationHub Data classes</a></li>
<li class="chapter" data-level="1.4.43" data-path="MacDonald_Annotation.html"><a href="#annotationhub-species"><i class="fa fa-check"></i><b>1.4.43</b> AnnotationHub Species</a></li>
<li class="chapter" data-level="1.4.44" data-path="MacDonald_Annotation.html"><a href="#annotationhub-data-sources"><i class="fa fa-check"></i><b>1.4.44</b> AnnotationHub Data sources</a></li>
<li class="chapter" data-level="1.4.45" data-path="MacDonald_Annotation.html"><a href="#annotationhub-query"><i class="fa fa-check"></i><b>1.4.45</b> AnnotationHub query</a></li>
<li class="chapter" data-level="1.4.46" data-path="MacDonald_Annotation.html"><a href="#annotationhub-query-1"><i class="fa fa-check"></i><b>1.4.46</b> AnnotationHub query</a></li>
<li class="chapter" data-level="1.4.47" data-path="MacDonald_Annotation.html"><a href="#selecting-annotationhub-resource"><i class="fa fa-check"></i><b>1.4.47</b> Selecting AnnotationHub resource</a></li>
<li class="chapter" data-level="1.4.48" data-path="MacDonald_Annotation.html"><a href="#annotationhub-exercises"><i class="fa fa-check"></i><b>1.4.48</b> AnnotationHub exercises</a></li>
<li class="chapter" data-level="1.4.49" data-path="MacDonald_Annotation.html"><a href="#biomart"><i class="fa fa-check"></i><b>1.4.49</b> biomaRt</a></li>
<li class="chapter" data-level="1.4.50" data-path="MacDonald_Annotation.html"><a href="#biomart-data-sets"><i class="fa fa-check"></i><b>1.4.50</b> biomaRt data sets</a></li>
<li class="chapter" data-level="1.4.51" data-path="MacDonald_Annotation.html"><a href="#biomart-queries"><i class="fa fa-check"></i><b>1.4.51</b> biomaRt queries</a></li>
<li class="chapter" data-level="1.4.52" data-path="MacDonald_Annotation.html"><a href="#biomart-attributes-and-filters"><i class="fa fa-check"></i><b>1.4.52</b> biomaRt attributes and filters</a></li>
<li class="chapter" data-level="1.4.53" data-path="MacDonald_Annotation.html"><a href="#biomart-query"><i class="fa fa-check"></i><b>1.4.53</b> biomaRt query</a></li>
<li class="chapter" data-level="1.4.54" data-path="MacDonald_Annotation.html"><a href="#biomart-exercises"><i class="fa fa-check"></i><b>1.4.54</b> biomaRt exercises</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:title:end-->
<!--bookdown:title:start-->
<div id="introduction-to-bioconductor-annotation-resources" class="section level1">
<h1><span class="header-section-number">1</span> Introduction to Bioconductor annotation resources</h1>
<div id="instructors" class="section level2">
<h2><span class="header-section-number">1.1</span> Instructors</h2>
<ul>
<li>James W. MacDonald (<a href="mailto:jmacdon@uw.edu">jmacdon@uw.edu</a>)</li>
<li>Lori Shepherd (<a href="mailto:lori.shepherd@roswellpark.org">lori.shepherd@roswellpark.org</a>)</li>
</ul>
</div>
<div id="workshop-description" class="section level2">
<h2><span class="header-section-number">1.2</span> Workshop Description</h2>
<p>There are various annotation packages provided by the Bioconductor project that can be used to incorporate additional information to results from high-throughput experiments. This can be as simple as mapping Ensembl IDs to corresponding HUGO gene symbols, to much more complex queries involving multiple data sources. In this workshop we will cover the various classes of annotation packages, what they contain, and how to use them efficiently.</p>
<div id="prerequisites" class="section level3">
<h3><span class="header-section-number">1.2.1</span> Prerequisites</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Basic understanding of the various annotation sources (NCBI, EBI/EMBL)</li>
</ul>
<p>Useful background reading</p>
<ul>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf">AnnotationDbi</a> vignette.</li>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html">biomaRt</a> vignette.</li>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf">GenomicFeatures</a> vignette.</li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3><span class="header-section-number">1.2.2</span> Workshop Participation</h3>
<p>After each type of annotation package is introduced, students will be given the opportunity to practice making their own queries.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3><span class="header-section-number">1.2.3</span> <em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li>AnnotationDbi</li>
<li>AnnotationHub</li>
<li>BSgenome</li>
<li>biomaRt</li>
<li>ensembldb</li>
<li>org.Hs.eg.db</li>
<li>TxDb.Hsapiens.UCSC.hg19.knownGene</li>
<li>EnsDb.Hsapiens.v79</li>
<li>EnsDb.Mmusculus.v79</li>
<li>Homo.sapiens</li>
<li>BSgenome.Hsapiens.UCSC.hg19</li>
<li>hugene20sttranscriptcluster.db</li>
</ul>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2><span class="header-section-number">1.3</span> Workshop goals and objectives</h2>
<p>Annotating data is a complex task. For any high-throughput experiment the analyst usually starts with a set of identifiers for each thing that was measured, and in order to make the results useful to collaborators these identifiers need to be mapped to other identifiers that are either more familiar to collaborators, or that can be used for further analyses. As an example, RNA-Seq data may only have Entrez Gene IDs for each gene measured, and as part of the output you may want to include the gene symbols, which are more likely to be familiar to a Biologist.</p>
<div id="learning-goals" class="section level3">
<h3><span class="header-section-number">1.3.1</span> Learning goals</h3>
<ul>
<li>Understand what sort of annotation data are available</li>
<li>Understand the difference between annotation sources (NCBI and EBI/EMBL)</li>
<li>Gain familiarity with the various ways to query annotation packages</li>
<li>Get some practice making queries</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3><span class="header-section-number">1.3.2</span> Learning objectives</h3>
<ul>
<li>Be able to use select and mapIds to map between identifiers</li>
<li>Be able to extract data from TxDb and EnsDb packages</li>
<li>Be able to make queries using biomaRt</li>
<li>Extract and utilize various data from AnnotationHub</li>
</ul>
</div>
</div>
<div id="annotation-workshop" class="section level2">
<h2><span class="header-section-number">1.4</span> Annotation Workshop</h2>
<div id="goals-for-this-workshop" class="section level3">
<h3><span class="header-section-number">1.4.1</span> Goals for this workshop</h3>
<ul>
<li><p>Learn about various annotation package types</p></li>
<li><p>Learn the basics of querying these resources</p></li>
<li><p>Discuss annotations in regard to Bioc data structures</p></li>
<li><p>Get in some practice</p></li>
</ul>
</div>
<div id="what-do-we-mean-by-annotation" class="section level3">
<h3><span class="header-section-number">1.4.2</span> What do we mean by annotation?</h3>
<p>Map a known ID to other functional or positional information</p>
<p><img src=".MacDonald_Annotation/mapping.png" style="width: 800px;"/></p>
</div>
<div id="specific-goal" class="section level3">
<h3><span class="header-section-number">1.4.3</span> Specific goal</h3>
<p>We have data and statistics, and we want to add other useful information</p>
<p><img src=".MacDonald_Annotation/boxes.png", style="width: 800px;"/></p>
<p>The end result might be as simple as a data.frame or HTML table, or as complex as a <code>RangedSummarizedExperiment</code></p>
</div>
<div id="data-containers" class="section level3">
<h3><span class="header-section-number">1.4.4</span> Data containers</h3>
<p><img src=".MacDonald_Annotation/eset.png", style="width: 800px;"/></p>
</div>
<div id="expressionset" class="section level3">
<h3><span class="header-section-number">1.4.5</span> ExpressionSet</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/eset.Rdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;Bioc2018Anno&quot;</span>))
eset</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 33552 features, 6 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM2194079 GSM2194080 ... GSM2194084 (6 total)
##   varLabels: title characteristics_ch1.1
##   varMetadata: labelDescription
## featureData
##   featureNames: 16657436 16657440 ... 17118478 (33552 total)
##   fvarLabels: PROBEID ENTREZID SYMBOL GENENAME
##   fvarMetadata: labelDescription
## experimentData: use &#39;experimentData(object)&#39;
## Annotation: pd.hugene.2.0.st</code></pre>
</div>
<div id="expressionset-continued" class="section level3">
<h3><span class="header-section-number">1.4.6</span> ExpressionSet (continued)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">exprs</span>(eset))</code></pre></div>
<pre><code>## [[1]]
## eset</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">pData</span>(<span class="kw">phenoData</span>(eset)))</code></pre></div>
<pre><code>##                          title characteristics_ch1.1
## GSM2194079   SW620-miR625-rep1     shRNA: miR-625-3p
## GSM2194080   SW620-miR625-rep2     shRNA: miR-625-3p
## GSM2194081   SW620-miR625-rep3     shRNA: miR-625-3p
## GSM2194082 SW620-scramble-rep1       shRNA: scramble
## GSM2194083 SW620-scramble-rep2       shRNA: scramble
## GSM2194084 SW620-scramble-rep3       shRNA: scramble</code></pre>
</div>
<div id="expressionset-continued-1" class="section level3">
<h3><span class="header-section-number">1.4.7</span> ExpressionSet (continued)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">pData</span>(<span class="kw">featureData</span>(eset)))</code></pre></div>
<pre><code>##           PROBEID  ENTREZID      SYMBOL
## 16657436 16657436     84771     DDX11L2
## 16657440 16657440 100302278   MIR1302-2
## 16657450 16657450    402483   LINC01000
## 16657469 16657469    140849 LINC00266-1
## 16657473 16657473    729759      OR4F29
## 16657476 16657476    388574   RPL23AP87
##                                                   GENENAME
## 16657436                     DEAD/H-box helicase 11 like 2
## 16657440                                   microRNA 1302-2
## 16657450       long intergenic non-protein coding RNA 1000
## 16657469      long intergenic non-protein coding RNA 266-1
## 16657473 olfactory receptor family 4 subfamily F member 29
## 16657476              ribosomal protein L23a pseudogene 87</code></pre>
</div>
<div id="bioc-containers-vs-basic-structures" class="section level3">
<h3><span class="header-section-number">1.4.8</span> BioC containers vs basic structures</h3>
<div id="pros" class="section level4">
<h4><span class="header-section-number">1.4.8.1</span> Pros</h4>
<ul>
<li><p>Validity checking</p></li>
<li><p>Subsetting</p></li>
<li><p>Function dispatch</p></li>
<li><p>Automatic behaviors</p></li>
</ul>
</div>
<div id="cons" class="section level4">
<h4><span class="header-section-number">1.4.8.2</span> Cons</h4>
<ul>
<li><p>Difficult to create</p></li>
<li><p>Cumbersome to extract data by hand</p></li>
<li><p>Useful only within R</p></li>
</ul>
</div>
</div>
<div id="annotation-sources" class="section level3">
<h3><span class="header-section-number">1.4.9</span> Annotation sources</h3>
<table>
<thead>
<tr class="header">
<th align="left">Package type</th>
<th align="left">Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ChipDb</td>
<td align="left">hugene20sttranscriptcluster.db</td>
</tr>
<tr class="even">
<td align="left">OrgDb</td>
<td align="left">org.Hs.eg.db</td>
</tr>
<tr class="odd">
<td align="left">TxDb/EnsDb</td>
<td align="left">TxDb.Hsapiens.UCSC.hg19.knownGene; EnsDb.Hsapiens.v75</td>
</tr>
<tr class="even">
<td align="left">OrganismDb</td>
<td align="left">Homo.sapiens</td>
</tr>
<tr class="odd">
<td align="left">BSgenome</td>
<td align="left">BSgenome.Hsapiens.UCSC.hg19</td>
</tr>
<tr class="even">
<td align="left">Others</td>
<td align="left">GO.db; KEGG.db</td>
</tr>
<tr class="odd">
<td align="left">AnnotationHub</td>
<td align="left">Online resource</td>
</tr>
<tr class="even">
<td align="left">biomaRt</td>
<td align="left">Online resource</td>
</tr>
</tbody>
</table>
</div>
<div id="interacting-with-annodb-packages" class="section level3">
<h3><span class="header-section-number">1.4.10</span> Interacting with AnnoDb packages</h3>
<p>The main function is <code>select</code>:</p>
<p>select(<em>annopkg</em>, <em>keys</em>, <em>columns</em>, <em>keytype</em>)</p>
<p>Where</p>
<ul>
<li><p>annopkg is the annotation package</p></li>
<li><p>keys are the IDs that we <strong>know</strong></p></li>
<li><p>columns are the values we <strong>want</strong></p></li>
<li>keytype is the type of key used
<ul>
<li>if the keytype is the <strong>central</strong> key, it can remain unspecified</li>
</ul></li>
</ul>
</div>
<div id="simple-example" class="section level3">
<h3><span class="header-section-number">1.4.11</span> Simple example</h3>
<p>Say we have analyzed data from an Affymetrix Human Gene ST 2.0 array and want to know what the genes are. For purposes of this lab, we just select some IDs at random.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(hugene20sttranscriptcluster.db)
<span class="kw">set.seed</span>(<span class="dv">12345</span>)
ids &lt;-<span class="st"> </span><span class="kw">featureNames</span>(eset)[<span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">25000</span>, <span class="dv">5</span>)]
ids</code></pre></div>
<pre><code>## [1] &quot;16908472&quot; &quot;16962185&quot; &quot;16920686&quot; &quot;16965513&quot; &quot;16819952&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">select</span>(hugene20sttranscriptcluster.db, ids, <span class="st">&quot;SYMBOL&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>##    PROBEID    SYMBOL
## 1 16908472 LINC01494
## 2 16962185      ALG3
## 3 16920686      &lt;NA&gt;
## 4 16965513      &lt;NA&gt;
## 5 16819952      CBFB</code></pre>
</div>
<div id="questions" class="section level3">
<h3><span class="header-section-number">1.4.12</span> Questions!</h3>
<p>How do you know what the central keys are?</p>
<ul>
<li><p>If it’s a ChipDb, the central key are the manufacturer’s probe IDs</p></li>
<li><p>It’s sometimes in the name - org.Hs.eg.db, where ‘eg’ means Entrez Gene ID</p></li>
<li><p>You can see examples using e.g., head(keys(<em>annopkg</em>)), and infer from that</p></li>
<li><p>But note that it’s never necessary to know the central key, as long as you specify the keytype</p></li>
</ul>
</div>
<div id="more-questions" class="section level3">
<h3><span class="header-section-number">1.4.13</span> More questions!</h3>
<p>What keytypes or columns are available for a given annotation package?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">keytypes</span>(hugene20sttranscriptcluster.db)</code></pre></div>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot; 
##  [5] &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;    
##  [9] &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;       
## [13] &quot;IPI&quot;          &quot;MAP&quot;          &quot;OMIM&quot;         &quot;ONTOLOGY&quot;    
## [17] &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;         &quot;PMID&quot;        
## [21] &quot;PROBEID&quot;      &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;      
## [25] &quot;UCSCKG&quot;       &quot;UNIGENE&quot;      &quot;UNIPROT&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">columns</span>(hugene20sttranscriptcluster.db)</code></pre></div>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot; 
##  [5] &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;    
##  [9] &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;       
## [13] &quot;IPI&quot;          &quot;MAP&quot;          &quot;OMIM&quot;         &quot;ONTOLOGY&quot;    
## [17] &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;         &quot;PMID&quot;        
## [21] &quot;PROBEID&quot;      &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;      
## [25] &quot;UCSCKG&quot;       &quot;UNIGENE&quot;      &quot;UNIPROT&quot;</code></pre>
</div>
<div id="another-example" class="section level3">
<h3><span class="header-section-number">1.4.14</span> Another example</h3>
<p>There is one issue with <code>select</code> however.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;16737401&#39;</span>,<span class="st">&#39;16657436&#39;</span> ,<span class="st">&#39;16678303&#39;</span>)
<span class="kw">select</span>(hugene20sttranscriptcluster.db, ids, <span class="kw">c</span>(<span class="st">&quot;SYMBOL&quot;</span>,<span class="st">&quot;MAP&quot;</span>))</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>##     PROBEID       SYMBOL     MAP
## 1  16737401        TRAF6   11p12
## 2  16657436      DDX11L1 1p36.33
## 3  16657436 LOC102725121 1p36.33
## 4  16657436      DDX11L2  2q14.1
## 5  16657436      DDX11L9 15q26.3
## 6  16657436     DDX11L10 16p13.3
## 7  16657436      DDX11L5  9p24.3
## 8  16657436     DDX11L16    Xq28
## 9  16657436     DDX11L16    Yq12
## 10 16678303         ARF1 1q42.13</code></pre>
</div>
<div id="the-mapids-function" class="section level3">
<h3><span class="header-section-number">1.4.15</span> The <code>mapIds</code> function</h3>
<p>An alternative to <code>select</code> is <code>mapIds</code>, which gives control of duplicates</p>
<ul>
<li><p>Same arguments as <code>select</code> with slight differences</p>
<ul>
<li><p>The columns argument can only specify one column</p></li>
<li><p>The keytype argument <strong>must</strong> be specified</p></li>
<li><p>An additional argument, multiVals used to control duplicates</p></li>
</ul></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mapIds</span>(hugene20sttranscriptcluster.db, ids, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;PROBEID&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>##  16737401  16657436  16678303 
##   &quot;TRAF6&quot; &quot;DDX11L1&quot;    &quot;ARF1&quot;</code></pre>
</div>
<div id="choices-for-multivals" class="section level3">
<h3><span class="header-section-number">1.4.16</span> Choices for multiVals</h3>
<p>Default is <code>first</code>, where we just choose the first of the duplicates. Other choices are <code>list</code>, <code>CharacterList</code>, <code>filter</code>, <code>asNA</code> or a user-specified function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mapIds</span>(hugene20sttranscriptcluster.db, ids, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;PROBEID&quot;</span>, <span class="dt">multiVals =</span> <span class="st">&quot;list&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>## $`16737401`
## [1] &quot;TRAF6&quot;
## 
## $`16657436`
## [1] &quot;DDX11L1&quot;      &quot;LOC102725121&quot; &quot;DDX11L2&quot;      &quot;DDX11L9&quot;     
## [5] &quot;DDX11L10&quot;     &quot;DDX11L5&quot;      &quot;DDX11L16&quot;    
## 
## $`16678303`
## [1] &quot;ARF1&quot;</code></pre>
</div>
<div id="choices-for-multivals-continued" class="section level3">
<h3><span class="header-section-number">1.4.17</span> Choices for multiVals (continued)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mapIds</span>(hugene20sttranscriptcluster.db, ids, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;PROBEID&quot;</span>, <span class="dt">multiVals =</span> <span class="st">&quot;CharacterList&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>## CharacterList of length 3
## [[&quot;16737401&quot;]] TRAF6
## [[&quot;16657436&quot;]] DDX11L1 LOC102725121 DDX11L2 DDX11L9 DDX11L10 DDX11L5 DDX11L16
## [[&quot;16678303&quot;]] ARF1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mapIds</span>(hugene20sttranscriptcluster.db, ids, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;PROBEID&quot;</span>, <span class="dt">multiVals =</span> <span class="st">&quot;filter&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>## 16737401 16678303 
##  &quot;TRAF6&quot;   &quot;ARF1&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mapIds</span>(hugene20sttranscriptcluster.db, ids, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;PROBEID&quot;</span>, <span class="dt">multiVals =</span> <span class="st">&quot;asNA&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>## 16737401 16657436 16678303 
##  &quot;TRAF6&quot;       NA   &quot;ARF1&quot;</code></pre>
</div>
<div id="chipdborgdb-questions" class="section level3">
<h3><span class="header-section-number">1.4.18</span> ChipDb/OrgDb questions</h3>
<p>Using either the hugene20sttranscriptcluster.db or org.Hs.eg.db package,</p>
<ul>
<li><p>What gene symbol corresponds to Entrez Gene ID 1000?</p></li>
<li><p>What is the Ensembl Gene ID for PPARG?</p></li>
<li><p>What is the UniProt ID for GAPDH?</p></li>
<li><p>How many of the probesets from the ExpressionSet (eset) we loaded map to a single gene? How many don’t map to a gene at all?</p></li>
</ul>
</div>
<div id="txdb-packages" class="section level3">
<h3><span class="header-section-number">1.4.19</span> TxDb packages</h3>
<p>TxDb packages contain positional information; the contents can be inferred by the package name</p>
<p><strong>TxDb.Species.Source.Build.Table</strong></p>
<ul>
<li><p>TxDb.Hsapiens.UCSC.hg19.knownGene</p>
<ul>
<li><p><em>Homo sapiens</em></p></li>
<li><p>UCSC genome browser</p></li>
<li><p>hg19 (their version of GRCh37)</p></li>
<li><p>knownGene table</p></li>
</ul></li>
</ul>
<p>TxDb.Dmelanogaster.UCSC.dm3.ensGene TxDb.Athaliana.BioMart.plantsmart22</p>
</div>
<div id="ensdb-packages" class="section level3">
<h3><span class="header-section-number">1.4.20</span> EnsDb packages</h3>
<p>EnsDb packages are similar to TxDb packages, but based on Ensembl mappings</p>
<p>EnsDb.Hsapiens.v79 EnsDb.Mmusculus.v79 EnsDb.Rnorvegicus.v79</p>
</div>
<div id="transcript-packages" class="section level3">
<h3><span class="header-section-number">1.4.21</span> Transcript packages</h3>
<p>As with ChipDb and OrgDb packages, <code>select</code> and <code>mapIds</code> can be used to make queries</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">select</span>(TxDb.Hsapiens.UCSC.hg19.knownGene, <span class="kw">c</span>(<span class="st">&quot;1&quot;</span>,<span class="st">&quot;10&quot;</span>),
       <span class="kw">c</span>(<span class="st">&quot;TXNAME&quot;</span>,<span class="st">&quot;TXCHROM&quot;</span>,<span class="st">&quot;TXSTART&quot;</span>,<span class="st">&quot;TXEND&quot;</span>), <span class="st">&quot;GENEID&quot;</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>##   GENEID     TXNAME TXCHROM  TXSTART    TXEND
## 1      1 uc002qsd.4   chr19 58858172 58864865
## 2      1 uc002qsf.2   chr19 58859832 58874214
## 3     10 uc003wyw.1    chr8 18248755 18258723</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">select</span>(EnsDb.Hsapiens.v79, <span class="kw">c</span>(<span class="st">&quot;1&quot;</span>, <span class="st">&quot;10&quot;</span>),
       <span class="kw">c</span>(<span class="st">&quot;GENEID&quot;</span>,<span class="st">&quot;GENENAME&quot;</span>,<span class="st">&quot;SEQNAME&quot;</span>,<span class="st">&quot;GENESEQSTART&quot;</span>,<span class="st">&quot;GENESEQEND&quot;</span>), <span class="st">&quot;ENTREZID&quot;</span>)</code></pre></div>
<pre><code>##   ENTREZID          GENEID GENENAME SEQNAME GENESEQSTART GENESEQEND
## 1        1 ENSG00000121410     A1BG      19     58345178   58353499
## 2       10 ENSG00000156006     NAT2       8     18391245   18401218</code></pre>
<p>But this is not how one normally uses them…</p>
</div>
<div id="granges" class="section level3">
<h3><span class="header-section-number">1.4.22</span> GRanges</h3>
<p>The normal use case for transcript packages is to extract positional information into a <code>GRanges</code> or <code>GRangesList</code> object. An example is the genomic position of all genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gns &lt;-<span class="st"> </span><span class="kw">genes</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)
gns</code></pre></div>
<pre><code>## GRanges object with 23056 ranges and 1 metadata column:
##         seqnames              ranges strand |     gene_id
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19   58858172-58874214      - |           1
##      10     chr8   18248755-18258723      + |          10
##     100    chr20   43248163-43280376      - |         100
##    1000    chr18   25530930-25757445      - |        1000
##   10000     chr1 243651535-244006886      - |       10000
##     ...      ...                 ...    ... .         ...
##    9991     chr9 114979995-115095944      - |        9991
##    9992    chr21   35736323-35743440      + |        9992
##    9993    chr22   19023795-19109967      - |        9993
##    9994     chr6   90539619-90584155      + |        9994
##    9997    chr22   50961997-50964905      - |        9997
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="grangeslist" class="section level3">
<h3><span class="header-section-number">1.4.23</span> GRangesList</h3>
<p>Or the genomic position of all transcripts <strong>by</strong> gene:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">txs &lt;-<span class="st"> </span><span class="kw">transcriptsBy</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)
txs</code></pre></div>
<pre><code>## GRangesList object of length 23459:
## $1 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58858172-58864865      - |     70455  uc002qsd.4
##   [2]    chr19 58859832-58874214      - |     70456  uc002qsf.2
## 
## $10 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand | tx_id    tx_name
##   [1]     chr8 18248755-18258723      + | 31944 uc003wyw.1
## 
## $100 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand | tx_id    tx_name
##   [1]    chr20 43248163-43280376      - | 72132 uc002xmj.3
## 
## ...
## &lt;23456 more elements&gt;
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="other-accessors" class="section level3">
<h3><span class="header-section-number">1.4.24</span> Other accessors</h3>
<ul>
<li><p>Positional information can be extracted for <code>transcripts</code>, <code>genes</code>, coding sequences (<code>cds</code>), <code>promoters</code> and <code>exons</code>.</p></li>
<li><p>Positional information can be extracted for most of the above, grouped by a second element. For example, our <code>transcriptsBy</code> call was all transcripts, grouped by gene.</p></li>
<li><p>More detail on these *Ranges objects is beyond the scope of this workshop, but why we want them is not.</p></li>
</ul>
</div>
<div id="why-ranges-objects" class="section level3">
<h3><span class="header-section-number">1.4.25</span> Why *Ranges objects</h3>
<p>The main rationale for *Ranges objects is to allow us to easily select and subset data based on genomic position information. This is really powerful!</p>
<p><code>GRanges</code> and <code>GRangesLists</code> act like data.frames and lists, and can be subsetted using the <code>[</code> function. As a really artificial example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">txs[txs <span class="op">%over%</span><span class="st"> </span>gns[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,]]</code></pre></div>
<pre><code>## GRangesList object of length 3:
## $1 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58858172-58864865      - |     70455  uc002qsd.4
##   [2]    chr19 58859832-58874214      - |     70456  uc002qsf.2
## 
## $10 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand | tx_id    tx_name
##   [1]     chr8 18248755-18258723      + | 31944 uc003wyw.1
## 
## $162968 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand | tx_id    tx_name
##   [1]    chr19 58865723-58874214      - | 70457 uc002qsh.2
##   [2]    chr19 58865723-58874214      - | 70458 uc002qsi.2
## 
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="ranges-use-cases" class="section level3">
<h3><span class="header-section-number">1.4.26</span> *Ranges use cases</h3>
<ul>
<li><p>Gene expression changes near differentially methylated CpG islands</p></li>
<li><p>Closest genes to a set of interesting SNPs</p></li>
<li><p>Genes near DNAseI hypersensitivity clusters</p></li>
<li><p>Number of CpGs measured over Gene X by Chip Y</p></li>
</ul>
</div>
<div id="summarizedexperiment-objects" class="section level3">
<h3><span class="header-section-number">1.4.27</span> SummarizedExperiment objects</h3>
<p>SummarizedExperiment objects are like ExpressionSets, but the row-wise annotations are GRanges, so you can subset by genomic locations:</p>
<p><img src=".MacDonald_Annotation/subset.png", style="width: 800px;"/></p>
<p>SummarizedExperiment objects are popular objects for representing expression data and other rectangular data (feature x sample data). Incoming packages are now strongly recommended to use this class representation instead of ExpressionSet.</p>
</div>
<div id="txdb-exercises" class="section level3">
<h3><span class="header-section-number">1.4.28</span> TxDb exercises</h3>
<ul>
<li><p>How many transcripts does PPARG have, according to UCSC?</p></li>
<li><p>Does Ensembl agree?</p></li>
<li>How many genes are between 2858473 and 3271812 on chr2 in the hg19 genome?
<ul>
<li>Hint: you make a <code>GRanges</code> like this - <code>GRanges(&quot;chr2&quot;, IRanges(2858473,3271812))</code></li>
</ul></li>
</ul>
</div>
<div id="organismdb-packages" class="section level3">
<h3><span class="header-section-number">1.4.29</span> OrganismDb packages</h3>
<p>OrganismDb packages are meta-packages that contain an OrgDb, a TxDb, and a GO.db package and allow cross-queries between those packages.</p>
<p>All previous accessors work; <code>select</code>, <code>mapIds</code>, <code>transcripts</code>, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Homo.sapiens)
Homo.sapiens</code></pre></div>
<pre><code>## OrganismDb Object:
## # Includes GODb Object:  GO.db 
## # With data about:  Gene Ontology 
## # Includes OrgDb Object:  org.Hs.eg.db 
## # Gene data about:  Homo sapiens 
## # Taxonomy Id:  9606 
## # Includes TxDb Object:  TxDb.Hsapiens.UCSC.hg19.knownGene 
## # Transcriptome data about:  Homo sapiens 
## # Based on genome:  hg19 
## # The OrgDb gene id ENTREZID is mapped to the TxDb gene id GENEID .</code></pre>
</div>
<div id="organismdb-packages-1" class="section level3">
<h3><span class="header-section-number">1.4.30</span> OrganismDb packages</h3>
<ul>
<li><p>Updateable - can change TxDb object</p></li>
<li><p>columns and keytypes span all underlying objects</p></li>
<li><p>Calls to TxDb accessors include a ‘columns’ argument</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">genes</span>(Homo.sapiens, <span class="dt">columns =</span> <span class="kw">c</span>(<span class="st">&quot;ENTREZID&quot;</span>,<span class="st">&quot;ALIAS&quot;</span>,<span class="st">&quot;UNIPROT&quot;</span>)),<span class="dv">4</span>)</code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>## GRanges object with 4 ranges and 3 metadata columns:
##        seqnames            ranges strand |                 ALIAS
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;CharacterList&gt;
##      1    chr19 58858172-58874214      - |       A1B,ABG,GAB,...
##     10     chr8 18248755-18258723      + |   AAC2,NAT-2,PNAT,...
##    100    chr20 43248163-43280376      - |                   ADA
##   1000    chr18 25530930-25757445      - | CD325,CDHN,CDw325,...
##                         UNIPROT     ENTREZID
##                 &lt;CharacterList&gt; &lt;FactorList&gt;
##      1            P04217,V9HWD8            1
##     10            A4Z6T7,P11245           10
##    100 A0A0S2Z381,P00813,F5GWI4          100
##   1000        P19022,A0A024RC42         1000
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="organismdb-exercises" class="section level3">
<h3><span class="header-section-number">1.4.31</span> OrganismDb exercises</h3>
<ul>
<li><p>Get all the GO terms for BRCA1</p></li>
<li><p>What gene does the UCSC transcript ID uc002fai.3 map to?</p></li>
<li><p>How many other transcripts does that gene have?</p></li>
<li><p>Get all the transcripts from the hg19 genome build, along with their Ensembl gene ID, UCSC transcript ID and gene symbol</p></li>
</ul>
</div>
<div id="organism.dplyr-package" class="section level3">
<h3><span class="header-section-number">1.4.32</span> Organism.dplyr package</h3>
<ul>
<li><p>Combines the data from TxDb and Org.Db associated packages into local database.</p></li>
<li>Allows functions from both <em>org.*</em> and <em>TxDb.*</em>
<ul>
<li><code>keytypes()</code>, <code>select()</code>, …</li>
<li><code>exons()</code>, <code>promoters()</code>, …</li>
</ul></li>
<li><p>Allows for filtering and display of combined TxDb and Org.Db information through <code>dplyr</code> functions.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Organism.dplyr)

<span class="co"># src = src_organism(&quot;TxDb.Hsapiens.UCSC.hg19.knownGene&quot;)</span>
src &lt;-<span class="st"> </span><span class="kw">src_organism</span>(<span class="dt">dbpath =</span> <span class="kw">hg38light</span>())
src</code></pre></div>
<pre><code>## src:  sqlite 3.22.0 [/home/lori/R/x86_64-pc-linux-gnu-library/3.5-BioC-3.8/Organism.dplyr/extdata/light.hg38.knownGene.sqlite]
## tbls: id, id_accession, id_go, id_go_all, id_omim_pm, id_protein,
##   id_transcript, ranges_cds, ranges_exon, ranges_gene, ranges_tx</code></pre>
</div>
<div id="organism.dply-example" class="section level3">
<h3><span class="header-section-number">1.4.33</span> Organism.dply example</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># TxTb function</span>
<span class="kw">promoters</span>(src)</code></pre></div>
<pre><code>## &lt;SQL&gt;
## SELECT *
## FROM `ranges_tx`</code></pre>
<pre><code>## GRanges object with 88 ranges and 2 metadata columns:
##                      seqnames              ranges strand |     tx_id
##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   uc001hzz.2             chr1 243843037-243845236      - |     15880
##   uc021plu.1             chr1 243843385-243845584      - |     15881
##   uc001iab.3             chr1 243843083-243845282      - |     15882
##   uc057qvr.1             chr1 243849929-243852128      - |     15883
##   uc057qvt.1             chr1 243614947-243617146      - |     15884
##          ...              ...                 ...    ... .       ...
##   uc064xqh.1 chrUn_GL000220v1       110025-112224      + |    197741
##   uc064xqi.1 chrUn_GL000220v1       112151-114350      + |    197742
##   uc064xqj.1 chrUn_GL000220v1       115428-117627      + |    197743
##   uc064xqk.1 chrUn_GL000220v1       116197-118396      + |    197744
##   uc033dnj.2 chrUn_GL000220v1       153997-156196      + |    197750
##                  tx_name
##              &lt;character&gt;
##   uc001hzz.2  uc001hzz.2
##   uc021plu.1  uc021plu.1
##   uc001iab.3  uc001iab.3
##   uc057qvr.1  uc057qvr.1
##   uc057qvt.1  uc057qvt.1
##          ...         ...
##   uc064xqh.1  uc064xqh.1
##   uc064xqi.1  uc064xqi.1
##   uc064xqj.1  uc064xqj.1
##   uc064xqk.1  uc064xqk.1
##   uc033dnj.2  uc033dnj.2
##   -------
##   seqinfo: 455 sequences (1 circular) from hg38 genome</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># see a table in database</span>
<span class="kw">tbl</span>(src, <span class="st">&quot;id&quot;</span>)</code></pre></div>
<pre><code>## # Source:   table&lt;id&gt; [?? x 6]
## # Database: sqlite 3.22.0 []
##    entrez map      ensembl         symbol genename               alias   
##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt;                  &lt;chr&gt;   
##  1 1      19q13.4  ENSG00000121410 A1BG   alpha-1-B glycoprotein A1B     
##  2 1      19q13.4  ENSG00000121410 A1BG   alpha-1-B glycoprotein ABG     
##  3 1      19q13.4  ENSG00000121410 A1BG   alpha-1-B glycoprotein GAB     
##  4 1      19q13.4  ENSG00000121410 A1BG   alpha-1-B glycoprotein HYST2477
##  5 1      19q13.4  ENSG00000121410 A1BG   alpha-1-B glycoprotein A1BG    
##  6 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  AAC2    
##  7 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  NAT-2   
##  8 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  PNAT    
##  9 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  NAT2    
## 10 100    20q13.12 ENSG00000196839 ADA    adenosine deaminase    ADA     
## # ... with more rows</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># information from two tables </span>
<span class="kw">inner_join</span>(<span class="kw">tbl</span>(src, <span class="st">&quot;id&quot;</span>), <span class="kw">tbl</span>(src, <span class="st">&quot;ranges_gene&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">            </span><span class="kw">filter</span>(symbol <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ADA&quot;</span>, <span class="st">&quot;NAT2&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">            </span>dplyr<span class="op">::</span><span class="kw">select</span>(gene_chrom, gene_start, gene_end,
            gene_strand, symbol, alias, map)</code></pre></div>
<pre><code>## Joining, by = &quot;entrez&quot;</code></pre>
<pre><code>## # Source:   lazy query [?? x 7]
## # Database: sqlite 3.22.0 []
##   gene_chrom gene_start gene_end gene_strand symbol alias map     
##   &lt;chr&gt;           &lt;int&gt;    &lt;int&gt; &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   
## 1 chr8         18391245 18401218 +           NAT2   AAC2  8p22    
## 2 chr8         18391245 18401218 +           NAT2   NAT-2 8p22    
## 3 chr8         18391245 18401218 +           NAT2   PNAT  8p22    
## 4 chr8         18391245 18401218 +           NAT2   NAT2  8p22    
## 5 chr20        44619522 44651742 -           ADA    ADA   20q13.12</code></pre>
</div>
<div id="organism.dplyr-exercises" class="section level3">
<h3><span class="header-section-number">1.4.34</span> Organism.dplyr exercises</h3>
<ul>
<li><p>How many supported organisms are implemented in Organism.dplyr?</p></li>
<li><p>Display the ensembl Id and genename description for symbol “NAT2”.</p></li>
<li><p>Show all the alias for “NAT2” in the database.</p></li>
<li><p>Display Gene ontology (GO) information for gene symbol “NAT2”.</p></li>
</ul>
</div>
<div id="bsgenome-packages" class="section level3">
<h3><span class="header-section-number">1.4.35</span> BSgenome packages</h3>
<p>BSgenome packages contain sequence information for a given species/build. There are many such packages - you can get a listing using <code>available.genomes</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BSgenome)
<span class="kw">head</span>(<span class="kw">available.genomes</span>())</code></pre></div>
<pre><code>## [1] &quot;BSgenome.Alyrata.JGI.v1&quot;                
## [2] &quot;BSgenome.Amellifera.BeeBase.assembly4&quot;  
## [3] &quot;BSgenome.Amellifera.UCSC.apiMel2&quot;       
## [4] &quot;BSgenome.Amellifera.UCSC.apiMel2.masked&quot;
## [5] &quot;BSgenome.Athaliana.TAIR.04232008&quot;       
## [6] &quot;BSgenome.Athaliana.TAIR.TAIR9&quot;</code></pre>
</div>
<div id="bsgenome-packages-1" class="section level3">
<h3><span class="header-section-number">1.4.36</span> BSgenome packages</h3>
<p>We can load and inspect a BSgenome package</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens</code></pre></div>
<pre><code>## Human genome:
## # organism: Homo sapiens (Human)
## # provider: UCSC
## # provider version: hg19
## # release date: Feb. 2009
## # release name: Genome Reference Consortium GRCh37
## # 93 sequences:
## #   chr1                  chr2                  chr3                 
## #   chr4                  chr5                  chr6                 
## #   chr7                  chr8                  chr9                 
## #   chr10                 chr11                 chr12                
## #   chr13                 chr14                 chr15                
## #   ...                   ...                   ...                  
## #   chrUn_gl000235        chrUn_gl000236        chrUn_gl000237       
## #   chrUn_gl000238        chrUn_gl000239        chrUn_gl000240       
## #   chrUn_gl000241        chrUn_gl000242        chrUn_gl000243       
## #   chrUn_gl000244        chrUn_gl000245        chrUn_gl000246       
## #   chrUn_gl000247        chrUn_gl000248        chrUn_gl000249       
## # (use &#39;seqnames()&#39; to see all the sequence names, use the &#39;$&#39; or &#39;[[&#39;
## # operator to access a given sequence)</code></pre>
</div>
<div id="bsgenome-packages-2" class="section level3">
<h3><span class="header-section-number">1.4.37</span> BSgenome packages</h3>
<p>The main accessor is <code>getSeq</code>, and you can get data by sequence (e.g., entire chromosome or unplaced scaffold), or by passing in a GRanges object, to get just a region.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getSeq</span>(Hsapiens, <span class="st">&quot;chr1&quot;</span>)</code></pre></div>
<pre><code>##   249250621-letter &quot;DNAString&quot; instance
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getSeq</span>(Hsapiens, gns[<span class="st">&quot;5467&quot;</span>,])</code></pre></div>
<pre><code>##   A DNAStringSet instance of length 1
##     width seq                                          names               
## [1] 85634 GCGGAGCGTGTGACGCTGCGG...TATTTAAGAGCTGACTGGAA 5467</code></pre>
<p>The Biostrings package contains most of the code for dealing with these <code>*StringSet</code> objects - please see the Biostrings vignettes and help pages for more information.</p>
</div>
<div id="bsgenome-exercises" class="section level3">
<h3><span class="header-section-number">1.4.38</span> BSgenome exercises</h3>
<ul>
<li>Get the sequences for all transcripts of the TP53 gene</li>
</ul>
</div>
<div id="annotationhub" class="section level3">
<h3><span class="header-section-number">1.4.39</span> AnnotationHub</h3>
<p>AnnotationHub is a package that allows us to query and download many different annotation objects, without having to explicitly install them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(AnnotationHub)
hub &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()</code></pre></div>
<pre><code>## snapshotDate(): 2018-06-27</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hub</code></pre></div>
<pre><code>## AnnotationHub with 44925 records
## # snapshotDate(): 2018-06-27 
## # $dataprovider: BroadInstitute, Ensembl, UCSC, ftp://ftp.ncbi.nlm.nih....
## # $species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos ta...
## # $rdataclass: GRanges, BigWigFile, FaFile, TwoBitFile, Rle, OrgDb, Cha...
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,
## #   tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH2&quot;]]&#39; 
## 
##             title                                                   
##   AH2     | Ailuropoda_melanoleuca.ailMel1.69.dna.toplevel.fa       
##   AH3     | Ailuropoda_melanoleuca.ailMel1.69.dna_rm.toplevel.fa    
##   AH4     | Ailuropoda_melanoleuca.ailMel1.69.dna_sm.toplevel.fa    
##   AH5     | Ailuropoda_melanoleuca.ailMel1.69.ncrna.fa              
##   AH6     | Ailuropoda_melanoleuca.ailMel1.69.pep.all.fa            
##   ...       ...                                                     
##   AH63655 | phastCons46wayPrimates.UCSC.hg19.chrX.rds               
##   AH63656 | phastCons46wayPrimates.UCSC.hg19.chrY.rds               
##   AH63657 | Alternative Splicing Annotation for Homo sapiens (Human)
##   AH63658 | Allele data from the IPD IMGT/HLA database              
##   AH63659 | Allele data from the IPD KIR database</code></pre>
</div>
<div id="querying-annotationhub" class="section level3">
<h3><span class="header-section-number">1.4.40</span> Querying AnnotationHub</h3>
<p>Finding the ‘right’ resource on AnnotationHub is like using Google - a well posed query is necessary to find what you are after. Useful queries are based on</p>
<ul>
<li><p>Data provider</p></li>
<li><p>Data class</p></li>
<li><p>Species</p></li>
<li><p>Data source</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(<span class="kw">mcols</span>(hub))</code></pre></div>
<pre><code>##  [1] &quot;title&quot;              &quot;dataprovider&quot;       &quot;species&quot;           
##  [4] &quot;taxonomyid&quot;         &quot;genome&quot;             &quot;description&quot;       
##  [7] &quot;coordinate_1_based&quot; &quot;maintainer&quot;         &quot;rdatadateadded&quot;    
## [10] &quot;preparerclass&quot;      &quot;tags&quot;               &quot;rdataclass&quot;        
## [13] &quot;rdatapath&quot;          &quot;sourceurl&quot;          &quot;sourcetype&quot;</code></pre>
</div>
<div id="annotationhub-data-providers" class="section level3">
<h3><span class="header-section-number">1.4.41</span> AnnotationHub Data providers</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(hub<span class="op">$</span>dataprovider)</code></pre></div>
<pre><code>##  [1] &quot;Ensembl&quot;                              
##  [2] &quot;UCSC&quot;                                 
##  [3] &quot;RefNet&quot;                               
##  [4] &quot;Inparanoid8&quot;                          
##  [5] &quot;NHLBI&quot;                                
##  [6] &quot;ChEA&quot;                                 
##  [7] &quot;Pazar&quot;                                
##  [8] &quot;NIH Pathway Interaction Database&quot;     
##  [9] &quot;Haemcode&quot;                             
## [10] &quot;BroadInstitute&quot;                       
## [11] &quot;PRIDE&quot;                                
## [12] &quot;Gencode&quot;                              
## [13] &quot;CRIBI&quot;                                
## [14] &quot;Genoscope&quot;                            
## [15] &quot;MISO, VAST-TOOLS, UCSC&quot;               
## [16] &quot;UWashington&quot;                          
## [17] &quot;Stanford&quot;                             
## [18] &quot;dbSNP&quot;                                
## [19] &quot;BioMart&quot;                              
## [20] &quot;GeneOntology&quot;                         
## [21] &quot;KEGG&quot;                                 
## [22] &quot;URGI&quot;                                 
## [23] &quot;ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/&quot;
## [24] &quot;EMBL-EBI&quot;</code></pre>
</div>
<div id="annotationhub-data-classes" class="section level3">
<h3><span class="header-section-number">1.4.42</span> AnnotationHub Data classes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(hub<span class="op">$</span>rdataclass)</code></pre></div>
<pre><code>##  [1] &quot;FaFile&quot;                           
##  [2] &quot;GRanges&quot;                          
##  [3] &quot;data.frame&quot;                       
##  [4] &quot;Inparanoid8Db&quot;                    
##  [5] &quot;TwoBitFile&quot;                       
##  [6] &quot;ChainFile&quot;                        
##  [7] &quot;SQLiteConnection&quot;                 
##  [8] &quot;biopax&quot;                           
##  [9] &quot;BigWigFile&quot;                       
## [10] &quot;AAStringSet&quot;                      
## [11] &quot;MSnSet&quot;                           
## [12] &quot;mzRpwiz&quot;                          
## [13] &quot;mzRident&quot;                         
## [14] &quot;list&quot;                             
## [15] &quot;TxDb&quot;                             
## [16] &quot;Rle&quot;                              
## [17] &quot;EnsDb&quot;                            
## [18] &quot;VcfFile&quot;                          
## [19] &quot;igraph&quot;                           
## [20] &quot;OrgDb&quot;                            
## [21] &quot;data.frame, DNAStringSet, GRanges&quot;</code></pre>
</div>
<div id="annotationhub-species" class="section level3">
<h3><span class="header-section-number">1.4.43</span> AnnotationHub Species</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">unique</span>(hub<span class="op">$</span>species))</code></pre></div>
<pre><code>## [1] &quot;Ailuropoda melanoleuca&quot; &quot;Anolis carolinensis&quot;   
## [3] &quot;Bos taurus&quot;             &quot;Caenorhabditis elegans&quot;
## [5] &quot;Callithrix jacchus&quot;     &quot;Canis familiaris&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(hub<span class="op">$</span>species))</code></pre></div>
<pre><code>## [1] 1971</code></pre>
</div>
<div id="annotationhub-data-sources" class="section level3">
<h3><span class="header-section-number">1.4.44</span> AnnotationHub Data sources</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(hub<span class="op">$</span>sourcetype)</code></pre></div>
<pre><code>##  [1] &quot;FASTA&quot;        &quot;UCSC track&quot;   &quot;GTF&quot;          &quot;TSV&quot;         
##  [5] &quot;Inparanoid&quot;   &quot;TwoBit&quot;       &quot;Chain&quot;        &quot;GRASP&quot;       
##  [9] &quot;Zip&quot;          &quot;CSV&quot;          &quot;BioPax&quot;       &quot;BioPaxLevel2&quot;
## [13] &quot;RData&quot;        &quot;BED&quot;          &quot;BigWig&quot;       &quot;tab&quot;         
## [17] &quot;mzTab&quot;        &quot;mzML&quot;         &quot;mzid&quot;         &quot;GFF&quot;         
## [21] &quot;ensembl&quot;      &quot;VCF&quot;          &quot;NCBI/ensembl&quot; &quot;NCBI/UniProt&quot;</code></pre>
</div>
<div id="annotationhub-query" class="section level3">
<h3><span class="header-section-number">1.4.45</span> AnnotationHub query</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">qry &lt;-<span class="st"> </span><span class="kw">query</span>(hub, <span class="kw">c</span>(<span class="st">&quot;granges&quot;</span>,<span class="st">&quot;homo sapiens&quot;</span>,<span class="st">&quot;ensembl&quot;</span>))
qry</code></pre></div>
<pre><code>## AnnotationHub with 56 records
## # snapshotDate(): 2018-06-27 
## # $dataprovider: Ensembl, UCSC
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,
## #   tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH5046&quot;]]&#39; 
## 
##             title                                          
##   AH5046  | Ensembl Genes                                  
##   AH5160  | Ensembl Genes                                  
##   AH5311  | Ensembl Genes                                  
##   AH5434  | Ensembl Genes                                  
##   AH5435  | Ensembl EST Genes                              
##   ...       ...                                            
##   AH60085 | Homo_sapiens.GRCh38.91.gtf                     
##   AH61125 | Homo_sapiens.GRCh38.92.abinitio.gtf            
##   AH61126 | Homo_sapiens.GRCh38.92.chr.gtf                 
##   AH61127 | Homo_sapiens.GRCh38.92.chr_patch_hapl_scaff.gtf
##   AH61128 | Homo_sapiens.GRCh38.92.gtf</code></pre>
</div>
<div id="annotationhub-query-1" class="section level3">
<h3><span class="header-section-number">1.4.46</span> AnnotationHub query</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">qry<span class="op">$</span>sourceurl</code></pre></div>
<pre><code>##  [1] &quot;rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg19/database/ensGene&quot;                                  
##  [2] &quot;rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg18/database/ensGene&quot;                                  
##  [3] &quot;rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg17/database/ensGene&quot;                                  
##  [4] &quot;rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg16/database/ensGene&quot;                                  
##  [5] &quot;rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg16/database/ensEstGene&quot;                               
##  [6] &quot;ftp://ftp.ensembl.org/pub/release-70/gtf/homo_sapiens/Homo_sapiens.GRCh37.70.gtf.gz&quot;                     
##  [7] &quot;ftp://ftp.ensembl.org/pub/release-69/gtf/homo_sapiens/Homo_sapiens.GRCh37.69.gtf.gz&quot;                     
##  [8] &quot;ftp://ftp.ensembl.org/pub/release-71/gtf/homo_sapiens/Homo_sapiens.GRCh37.71.gtf.gz&quot;                     
##  [9] &quot;ftp://ftp.ensembl.org/pub/release-72/gtf/homo_sapiens/Homo_sapiens.GRCh37.72.gtf.gz&quot;                     
## [10] &quot;ftp://ftp.ensembl.org/pub/release-73/gtf/homo_sapiens/Homo_sapiens.GRCh37.73.gtf.gz&quot;                     
## [11] &quot;ftp://ftp.ensembl.org/pub/release-74/gtf/homo_sapiens/Homo_sapiens.GRCh37.74.gtf.gz&quot;                     
## [12] &quot;ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz&quot;                     
## [13] &quot;ftp://ftp.ensembl.org/pub/release-78/gtf/homo_sapiens/Homo_sapiens.GRCh38.78.gtf.gz&quot;                     
## [14] &quot;ftp://ftp.ensembl.org/pub/release-76/gtf/homo_sapiens/Homo_sapiens.GRCh38.76.gtf.gz&quot;                     
## [15] &quot;ftp://ftp.ensembl.org/pub/release-79/gtf/homo_sapiens/Homo_sapiens.GRCh38.79.gtf.gz&quot;                     
## [16] &quot;ftp://ftp.ensembl.org/pub/release-77/gtf/homo_sapiens/Homo_sapiens.GRCh38.77.gtf.gz&quot;                     
## [17] &quot;ftp://ftp.ensembl.org/pub/release-80/gtf/homo_sapiens/Homo_sapiens.GRCh38.80.gtf.gz&quot;                     
## [18] &quot;ftp://ftp.ensembl.org/pub/release-81/gtf/homo_sapiens/Homo_sapiens.GRCh38.81.gtf.gz&quot;                     
## [19] &quot;ftp://ftp.ensembl.org/pub/release-82/gtf/homo_sapiens/Homo_sapiens.GRCh38.82.gtf.gz&quot;                     
## [20] &quot;ftp://ftp.ensembl.org/pub/release-83/gtf/homo_sapiens/Homo_sapiens.GRCh38.83.gtf.gz&quot;                     
## [21] &quot;ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.abinitio.gtf.gz&quot;            
## [22] &quot;ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.chr.gtf.gz&quot;                 
## [23] &quot;ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.chr_patch_hapl_scaff.gtf.gz&quot;
## [24] &quot;ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz&quot;                     
## [25] &quot;ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.abinitio.gtf.gz&quot;            
## [26] &quot;ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.chr.gtf.gz&quot;                 
## [27] &quot;ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.chr_patch_hapl_scaff.gtf.gz&quot;
## [28] &quot;ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.gtf.gz&quot;                     
## [29] &quot;ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.abinitio.gtf.gz&quot;            
## [30] &quot;ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.chr.gtf.gz&quot;                 
## [31] &quot;ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.chr_patch_hapl_scaff.gtf.gz&quot;
## [32] &quot;ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.gtf.gz&quot;                     
## [33] &quot;ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.abinitio.gtf.gz&quot;            
## [34] &quot;ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.chr.gtf.gz&quot;                 
## [35] &quot;ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.chr_patch_hapl_scaff.gtf.gz&quot;
## [36] &quot;ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.gtf.gz&quot;                     
## [37] &quot;ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.abinitio.gtf.gz&quot;            
## [38] &quot;ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.chr.gtf.gz&quot;                 
## [39] &quot;ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.chr_patch_hapl_scaff.gtf.gz&quot;
## [40] &quot;ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.gtf.gz&quot;                     
## [41] &quot;ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.abinitio.gtf.gz&quot;            
## [42] &quot;ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.chr.gtf.gz&quot;                 
## [43] &quot;ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.chr_patch_hapl_scaff.gtf.gz&quot;
## [44] &quot;ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.gtf.gz&quot;                     
## [45] &quot;ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.abinitio.gtf.gz&quot;            
## [46] &quot;ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.chr.gtf.gz&quot;                 
## [47] &quot;ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.chr_patch_hapl_scaff.gtf.gz&quot;
## [48] &quot;ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.gtf.gz&quot;                     
## [49] &quot;ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.abinitio.gtf.gz&quot;            
## [50] &quot;ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.chr.gtf.gz&quot;                 
## [51] &quot;ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.chr_patch_hapl_scaff.gtf.gz&quot;
## [52] &quot;ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz&quot;                     
## [53] &quot;ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.abinitio.gtf.gz&quot;            
## [54] &quot;ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.chr.gtf.gz&quot;                 
## [55] &quot;ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.chr_patch_hapl_scaff.gtf.gz&quot;
## [56] &quot;ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.gtf.gz&quot;</code></pre>
</div>
<div id="selecting-annotationhub-resource" class="section level3">
<h3><span class="header-section-number">1.4.47</span> Selecting AnnotationHub resource</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">whatIwant &lt;-<span class="st"> </span>qry[[<span class="st">&quot;AH50377&quot;</span>]]</code></pre></div>
<p>We can use these data as they are, or convert to a TxDb format:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GRCh38TxDb &lt;-<span class="st"> </span><span class="kw">makeTxDbFromGRanges</span>(whatIwant)
GRCh38TxDb</code></pre></div>
<pre><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Genome: GRCh38
## # transcript_nrow: 199184
## # exon_nrow: 675836
## # cds_nrow: 270225
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2018-06-29 10:02:16 -0400 (Fri, 29 Jun 2018)
## # GenomicFeatures version at creation time: 1.33.0
## # RSQLite version at creation time: 2.1.1
## # DBSCHEMAVERSION: 1.2</code></pre>
</div>
<div id="annotationhub-exercises" class="section level3">
<h3><span class="header-section-number">1.4.48</span> AnnotationHub exercises</h3>
<ul>
<li><p>How many resources are on AnnotationHub for Atlantic salmon (Salmo salar)?</p></li>
<li><p>Get the most recent Ensembl build for domesticated dog (Canis familiaris) and make a TxDb</p></li>
</ul>
</div>
<div id="biomart" class="section level3">
<h3><span class="header-section-number">1.4.49</span> biomaRt</h3>
<p>The biomaRt package allows queries to an Ensembl Biomart server. We can see the choices of servers that we can use:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(biomaRt)
<span class="kw">listMarts</span>()</code></pre></div>
<pre><code>##                biomart               version
## 1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 92
## 2   ENSEMBL_MART_MOUSE      Mouse strains 92
## 3     ENSEMBL_MART_SNP  Ensembl Variation 92
## 4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 92</code></pre>
</div>
<div id="biomart-data-sets" class="section level3">
<h3><span class="header-section-number">1.4.50</span> biomaRt data sets</h3>
<p>And we can then check for the available data sets on a particular server.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mart &lt;-<span class="st"> </span><span class="kw">useMart</span>(<span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>)
<span class="kw">head</span>(<span class="kw">listDatasets</span>(mart))</code></pre></div>
<pre><code>##                       dataset                        description
## 1  acarolinensis_gene_ensembl     Anole lizard genes (AnoCar2.0)
## 2   amelanoleuca_gene_ensembl              Panda genes (ailMel1)
## 3     amexicanus_gene_ensembl        Cave fish genes (AstMex102)
## 4     anancymaae_gene_ensembl Ma&#39;s night monkey genes (Anan_2.0)
## 5 aplatyrhynchos_gene_ensembl          Duck genes (BGI_duck_1.0)
## 6        btaurus_gene_ensembl                 Cow genes (UMD3.1)
##        version
## 1    AnoCar2.0
## 2      ailMel1
## 3    AstMex102
## 4     Anan_2.0
## 5 BGI_duck_1.0
## 6       UMD3.1</code></pre>
</div>
<div id="biomart-queries" class="section level3">
<h3><span class="header-section-number">1.4.51</span> biomaRt queries</h3>
<p>After setting up a <code>mart</code> object pointing to the server and data set that we care about, we can make queries. We first set up the <code>mart</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mart &lt;-<span class="st"> </span><span class="kw">useMart</span>(<span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>,<span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</code></pre></div>
<p>Queries are of the form</p>
<p>getBM(<em>attributes</em>, <em>filters</em>, <em>values</em>, <em>mart</em>)</p>
<p>where</p>
<ul>
<li><p>attributes are the things we <strong>want</strong></p></li>
<li><p>filters are the <em>types of</em> IDs we <strong>have</strong></p></li>
<li><p>values are the IDs we <strong>have</strong></p></li>
<li><p>mart is the <code>mart</code> object we set up</p></li>
</ul>
</div>
<div id="biomart-attributes-and-filters" class="section level3">
<h3><span class="header-section-number">1.4.52</span> biomaRt attributes and filters</h3>
<p>Both attributes and filters have rather inscrutable names, but a listing can be accessed using</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">atrib &lt;-<span class="st"> </span><span class="kw">listAttributes</span>(mart)
filts &lt;-<span class="st"> </span><span class="kw">listFilters</span>(mart)
<span class="kw">head</span>(atrib)</code></pre></div>
<pre><code>##                            name                  description         page
## 1               ensembl_gene_id               Gene stable ID feature_page
## 2       ensembl_gene_id_version       Gene stable ID version feature_page
## 3         ensembl_transcript_id         Transcript stable ID feature_page
## 4 ensembl_transcript_id_version Transcript stable ID version feature_page
## 5            ensembl_peptide_id            Protein stable ID feature_page
## 6    ensembl_peptide_id_version    Protein stable ID version feature_page</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(filts)</code></pre></div>
<pre><code>##              name              description
## 1 chromosome_name Chromosome/scaffold name
## 2           start                    Start
## 3             end                      End
## 4      band_start               Band Start
## 5        band_end                 Band End
## 6    marker_start             Marker Start</code></pre>
</div>
<div id="biomart-query" class="section level3">
<h3><span class="header-section-number">1.4.53</span> biomaRt query</h3>
<p>A simple example query</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">afyids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;1000_at&quot;</span>,<span class="st">&quot;1001_at&quot;</span>,<span class="st">&quot;1002_f_at&quot;</span>,<span class="st">&quot;1007_s_at&quot;</span>)
<span class="kw">getBM</span>(<span class="kw">c</span>(<span class="st">&quot;affy_hg_u95av2&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>), <span class="kw">c</span>(<span class="st">&quot;affy_hg_u95av2&quot;</span>), afyids, mart)</code></pre></div>
<pre><code>##   affy_hg_u95av2 hgnc_symbol
## 1        1000_at       MAPK3
## 2      1007_s_at        DDR1
## 3      1002_f_at            
## 4      1002_f_at     CYP2C19
## 5        1001_at        TIE1</code></pre>
</div>
<div id="biomart-exercises" class="section level3">
<h3><span class="header-section-number">1.4.54</span> biomaRt exercises</h3>
<ul>
<li><p>Get the Ensembl gene IDs and HUGO symbol for Entrez Gene IDs 672, 5468 and 7157</p></li>
<li><p>What do you get if you query for the ‘gene_exon’ for GAPDH?</p></li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
